myia/opt/dde.py

```diff
@@ -1,226 +1,479 @@
+    P.scalar_mul,
+    P.scalar_ne,
+    P.scalar_pow,
+    P.scalar_sign,
+    P.scalar_sin,
+    P.scalar_sub,
+    P.scalar_tan,
+    P.scalar_tanh,
+    P.scalar_to_array,
+    P.scalar_trunc,
+    P.scalar_uadd,
+    P.scalar_usub,
+    P.scatter,
+    P.scatter_add,
+    P.shape,
+    P.split,
+    P.transpose,
+)
+def _vprop_generic(vprop, need, inputs, out):
+    for inp in inputs:
+        vprop.add_need(inp, ANYTHING)
+    vprop.add_value(out, need, ANYTHING)
 
-        Returns {graph: {path: [(graph2, *path2), ...], ...}, ...}
 
-        Meaning that computing graph(...)[*path] may require computing
-        graph2(...)[*path2].
-        """
-        results = {}
-        paths = self.node_to_paths(root)
+#######################
+# DeadDataElimination #
+#######################
 
-        for g in root.manager.graphs:
-            results[g] = {}
-            for path, (parent, idx) in self.output_structure(g):
-                node = parent.inputs[idx]
-                all_paths = set()
-                for node2 in dfs(node, succ_incoming):
-                    all_paths.update(paths.get(node2, []))
-                results[g][path] = (parent, idx), all_paths
 
-        return results
+class DeadDataElimination(Partializable):
+    """Eliminate expressions that compute unretrieved data."""
 
-    def find_dead_paths(self, root):
-        """Returns a set of dead paths for each graph.
+    def __init__(self, resources=None):
+        """Initialize a DeadDataElimination."""
+        self.resources = resources
 
-        Returns {graph: {(node, idx) ...}} where the edge represented
-        by (node, idx) is dead.
-        """
-        deps = self.dependencies(root)
-
-        seen = set()
-        keep = set()
-
-        def succ(path):
-            if path in seen:
-                return
-            g, *path = path
-            if g not in deps:
-                return
-            path = tuple(path)
-            sg = deps[g]
-            while path and path not in sg:
-                path = path[:-1]
-            seen.add((g, *path))
-            keep.update(_subslices((g, *path)))
-            node, paths = sg[path]
-            for p in paths:
-                yield p
-
-        for path in deps.get(root, []):
-            for _ in dfs((root, *path), succ):
-                pass
-
-        missing = {}
-        for g, paths in deps.items():
-            missing[g] = set(node for (p, (node, _)) in paths.items()
-                             if (g, *p) not in keep
-                             and all(subp not in seen
-                                     for subp in _subslices((g, *p))))
-        return missing
+    def make_dead(self, node):
+        """Create a dead version of the node."""
+        a = node.abstract
+        val = DEAD
+        if isinstance(a, abstract.AbstractScalar):
+            if a.xtype() == xtype.EnvType:
+                val = newenv
+            assert a.xtype() != xtype.UniverseType  # TODO
```

Comment:
A workaround for now might be to just leave it alone.  It might be better than crashing.

Reply:
True, I could do that.